ostd/src/arch/riscv/irq/chip/plic.rs

```diff
@@ -0,0 +1,351 @@
+                    .property("riscv,ndev")
+                    .and_then(|prop| prop.as_usize())
+                    .expect("Failed to read 'riscv,ndev' property from PLIC node")
+                    as u32;
+
+                let hart_to_target_mapping = {
+                    let cpu_nodes = fdt
+                        .find_node("/cpus")
+                        .expect("Failed to find '/cpus' node")
+                        .children()
+                        .filter(|node| node.name.split('@').next().unwrap() == "cpu")
+                        .collect::<Vec<_>>();
+
+                    plic_node
+                        .property("interrupts-extended")
+                        .expect("Failed to read 'interrupts-extended' property from PLIC node")
+                        .value
+                        .chunks_exact(8)
+                        .enumerate()
+                        .filter_map(|(idx, chunk)| {
+                            let target = [
+                                u32::from_be_bytes(chunk[0..4].try_into().unwrap()),
+                                u32::from_be_bytes(chunk[4..8].try_into().unwrap()),
+                            ];
+
+                            if target[1] != 0x09 {
+                                return None;
+                            }
+
+                            let hart_id = cpu_nodes.iter().find_map(|node| {
+                                node.children()
+                                    .find(|child| {
+                                        child
+                                            .compatible()
+                                            .is_some_and(|c| c.all().any(|s| s == "riscv,cpu-intc"))
+                                            && child.property("phandle").is_some_and(|ph| {
+                                                ph.as_usize().unwrap() as u32 == target[0]
+                                            })
+                                    })
+                                    .and_then(|_| node.property("reg").and_then(|p| p.as_usize()))
+                            })?;
+
+                            Some((hart_id as u32, idx as u32))
+                        })
+                        .collect()
+                };
+
+                Self {
+                    phandle,
+                    io_mem: io_mem_builder.reserve(range),
+                    num_interrupt_sources,
+                    hart_to_target_mapping,
+                    interrupt_number_mappings: (0..num_interrupt_sources)
+                        .map(|_| None)
+                        .collect::<Vec<Option<u32>>>()
+                        .into_boxed_slice(),
+                }
+            })
+            .collect()
+    }
+
+    pub(super) fn map_interrupt_source_to(
+        &mut self,
+        interrupt_source: u32,
+        irq_line: &IrqLine,
+    ) -> Result<()> {
+        assert!(interrupt_source > 0 && interrupt_source < self.num_interrupt_sources);
```

Comment:
~~This assertion should be removed? `interrupt_number_mappings[interrupt_source as usize]` will check it.~~

Oh, this still rejects the case where `interrupt_source == 0`. Perhaps make it explicit.

Maybe we should not store the same information twice. We can add a method `num_interrupt_sources()` and remove the `num_interrupt_sources` field?
```rust
    fn num_interrupt_sources(&self) -> usize {
        self.interrupt_number_mappings.len()
    }
```